envision/web/src/components/components.js

```diff
@@ -0,0 +1,343 @@
+  }
+
+  const egoCamRootRef = useRef();
+  const thirdPersonCameraRef = useRef();
+
+  let canvas = scene.getEngine().getRenderingCanvas();
+
+  if (!egoCamRootRef.current) {
+    let egoCamRoot = new TransformNode("ego-camera-root");
+    egoCamRoot.position = new Vector3.Zero(); // Set to the ego vehicle position during update
+    let egoCamera = new UniversalCamera(
+      "ego-camera",
+      new Vector3(0, 5, -15), // Relative to camera root position
+      scene
+    );
+    egoCamera.parent = egoCamRoot;
+    egoCamRootRef.current = egoCamRoot;
+  }
+
+  if (!thirdPersonCameraRef.current) {
+    let thirdPersonCamera = new ArcRotateCamera(
+      "third-person-camera",
+      -Math.PI / 2, // alpha
+      0, // beta
+      200, // radius
+      new Vector3(0, 0, 0), // target
+      scene
+    );
+    thirdPersonCamera.attachControl(canvas, true);
+    thirdPersonCamera.panningSensibility = 50;
+    thirdPersonCamera.lowerRadiusLimit = 5;
+    scene.activeCamera = thirdPersonCamera; // default camera
+    thirdPersonCameraRef.current = thirdPersonCamera;
+  }
+
+  // Update third person camera's pointing target and radius
+  useEffect(() => {
+    if (roadNetworkBbox.length != 4) {
+      return;
+    }
+
+    let mapCenter = [
+      (roadNetworkBbox[0] + roadNetworkBbox[2]) / 2,
+      (roadNetworkBbox[1] + roadNetworkBbox[3]) / 2,
+    ];
+    let thirdPersonCamera = thirdPersonCameraRef.current;
+    thirdPersonCamera.target.x = mapCenter[0];
+    thirdPersonCamera.target.z = mapCenter[1];
+    thirdPersonCamera.radius = Math.max(
+      Math.abs(roadNetworkBbox[0] - roadNetworkBbox[2]),
+      Math.abs(roadNetworkBbox[1] - roadNetworkBbox[3])
+    );
+  }, [JSON.stringify(roadNetworkBbox)]);
+
+  // Set active camera
+  useEffect(() => {
+    let egoCamRoot = egoCamRootRef.current;
+    let thirdPersonCamera = thirdPersonCameraRef.current;
+
+    if (egoView) {
+      let egoCamera = egoCamRoot.getChildren()[0];
+      scene.activeCamera = egoCamera;
+
+      // Disable mouse input to the third person camera during ego view
+      thirdPersonCamera.detachControl(canvas);
+    } else {
+      thirdPersonCamera.attachControl(canvas, true);
+      scene.activeCamera = thirdPersonCamera;
+    }
+  }, [egoView]);
+
+  return null;
+}
+
+let meshesLoaded = (vehicleMeshTemplates) => {
+  for (const [_, mesh] of Object.entries(vehicleMeshTemplates)) {
+    if (mesh == null) {
+      return false;
+    }
+  }
+  return true;
+};
+
+// Vehicles and agent labels
+export function Vehicles({ scene, worldState, vehicleRootUrl, egoView }) {
+  if (scene == null) {
+    return null;
+  }
+
+  const vehicleMeshTemplatesRef = useRef({});
+  const vehicleMeshesRef = useRef({});
+  const agentLabelGeometryRef = useRef({});
+
+  let vehicleMeshTemplates = vehicleMeshTemplatesRef.current;
+  let vehicleMeshes = vehicleMeshesRef.current;
+  let agentLabelGeometry = agentLabelGeometryRef.current;
+
+  // Load mesh asynchronously
+  useEffect(() => {
```

Comment:
To improve readability, let's do a slight restructure of the code in this useEffect to leverage some methods.